src/main/java/shef/servlets/RecipeImageUploadUrlServlet.java

```diff
@@ -0,0 +1,82 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package shef.servlets;
+
+import com.google.appengine.api.blobstore.BlobInfo;
+import com.google.appengine.api.blobstore.BlobInfoFactory;
+import com.google.appengine.api.blobstore.BlobKey;
+import com.google.appengine.api.blobstore.BlobstoreService;
+import com.google.appengine.api.blobstore.BlobstoreServiceFactory;
+import com.google.appengine.api.images.ImagesService;
+import com.google.appengine.api.images.ImagesServiceFactory;
+import com.google.appengine.api.images.ServingUrlOptions;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.io.IOException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.util.Map;
+import java.util.List;
+
+@WebServlet("/recipe-image-upload-url")
+public class RecipeImageUploadUrlServlet extends HttpServlet {
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
+    String uploadUrl = blobstoreService.createUploadUrl("/new-recipe");
+    response.setContentType("text/html");
+    response.getWriter().println(uploadUrl);
+  }
+
+  /** Returns a URL that points to the uploaded file, or null if the user didn't upload a file. */
+  public static String getRecipeImageUrl(HttpServletRequest request, String formInputElementName) {
```

Comment:
I'm sorry, I should have warned you about this earlier, but it slipped my mind. When I worked with Blobstore before, I encountered an error that only happens in the deployed app. The solution is described here, in the ImageServiceFailureException section of the STEP Internship Known Issues doc. https://docs.google.com/document/d/1_Hyte0_3IJ9nPglQ4G116BXjuIK3dXxI5tQXy9rPpp0/edit#
Essentially, instead of getting and storing an url for the image, you must instead store a "Blob key". When you need the image, you need to serve it yourself using BlobServlet.java, which I have already created. You should look at the code I wrote for storing and retrieving the User's profile picture as a reference. And feel free to ask me for clarification

Reply:
Got it, thanks again for the heads up!